conwayste/src/ui/layer.rs

```diff
@@ -31,137 +31,251 @@ use super::{
+            return Ok(widget);
         }
 
-        let error_msg = format!("ERROR in Layer::enter_focus() =>  Widget (ID: {:?}) not found in layer (ID: {:?})",
-            id,
-            self.id
-        );
-        Err(Box::new(UIError::WidgetNotFound{reason: error_msg}))
+        // If it doesn't belong to a pane, check the widget list for an entry
+        self.widget_list
+            .iter_mut()
+            .filter(|widget| widget.id() == widget_id)
+            .next()
+            .ok_or_else(|| Box::new(UIError::WidgetNotFound {
+                reason: format!("{:?} not found in layering's widget list", widget_id).to_owned()
+            }))
     }
 
-    /// Iterates through all of the widgets grouped in this layer searching for the specified WidgetID.
-    /// If a Pane widget is found, it will search through all of its contained elements as well.
-    /// In either scenario, the first element found will be returned.
-    pub fn get_widget_mut(&mut self, id: WidgetID) -> UIResult<&mut Box<dyn Widget>>
-    {
-        let mut index = None;
-        let mut pane_index = None;
-
-        for (i, w) in self.widgets.iter().enumerate() {
-            if w.id() == id {
-                index = Some(i);
-                break;
-            }
+    /// Add a widget to the layering at the provided z_index depth. Internal data structure
+    /// maintains a list sorted by descending z_index.
+    ///
+    /// # Error
+    /// A WidgetIDCollision error can be returned if the widget-id exists in this layering.
+    pub fn add_widget(&mut self, widget: Box<dyn Widget>, z_index: usize) -> UIResult<()> {
+        if self.check_for_entry(widget.id()) {
+            return Err(Box::new(UIError::WidgetIDCollision {
+                    reason: format!("Widget with ID {:?} exists in layer's widget list.", widget.id())
+            }));
+        }
 
-            if let Some(pane) = w.downcast_ref::<Pane>() {
-                for (j, x) in pane.widgets.iter().enumerate() {
-                    if x.id() == id {
-                        index = Some(i);
-                        pane_index = Some(j);
-                    }
-                }
+        // Insert by z-index, descending. Use zero when the list is empty.
+        let insertion_index = self.widget_list
+            .iter()
+            .position(|widget| z_index >= widget.z_index())
+            .unwrap_or(0);
+
+        self.widget_list.insert(insertion_index, widget);
+        self.rebuild_id_cache();
+        Ok(())
+    }
+
+    /// Removes a widget belonging to the layering
+    ///
+    /// # Error
+    /// A WidgetNotFound error can be returned if a widget with the `widget_id` does not exist
+    /// in the internal list of widgets.
+    pub fn _remove_widget(&mut self, widget_id: WidgetID) -> UIResult<()> {
+        let removal_index = self.widget_list
+            .iter()
+            .position(|widget| widget_id >= widget.id())
```

Comment:
What is the reason this is `>=`?

Reply:
It's a bug! Great catch. Should be `==`.